cupyx/scipy/linalg/special_matrices.py

```diff
@@ -0,0 +1,552 @@
+    return a
+
+
+def kron(a, b):
+    """Kronecker product.
+
+    The result is the block matrix::
+        a[0,0]*b    a[0,1]*b  ... a[0,-1]*b
+        a[1,0]*b    a[1,1]*b  ... a[1,-1]*b
+        ...
+        a[-1,0]*b   a[-1,1]*b ... a[-1,-1]*b
+
+    Args:
+        a (cupy.ndarray): Input array
+        b (cupy.ndarray): Input array
+
+    Returns:
+        cupy.ndarray: Kronecker product of ``a`` and ``b``.
+
+    .. seealso:: :func:`scipy.linalg.kron`
+    """
+    o = cupy.outer(a, b)
+    o = o.reshape(a.shape + b.shape)
+    return cupy.concatenate(cupy.concatenate(o, axis=1), axis=1)
+
+
+def block_diag(*arrs):
+    """Create a block diagonal matrix from provided arrays.
+
+    Given the inputs ``A``, ``B`` and ``C``, the output will have these
+    arrays arranged on the diagonal::
+        [[A, 0, 0],
+         [0, B, 0],
+         [0, 0, C]]
+
+    Args:
+        A, B, C, ... (cupy.ndarray): Input arrays. A 1-D array of length ``n``
+            is treated as a 2-D array with shape ``(1,n)``.
+
+    Returns:
+        (cupy.ndarray): Array with ``A``, ``B``, ``C``, ... on the diagonal.
+            Output has the same dtype as ``A``.
+
+    .. seealso:: :func:`scipy.linalg.block_diag`
+    """
+    if not arrs:
+        return cupy.empty((1, 0))
+
+    # Convert to 2D and check
+    if len(arrs) == 1:
+        arrs = (cupy.atleast_2d(*arrs),)
+    else:
+        arrs = cupy.atleast_2d(*arrs)
+    if any(a.ndim != 2 for a in arrs):
+        bad = [k for k in range(len(arrs)) if arrs[k].ndim != 2]
+        raise ValueError("arguments in the following positions have dimension "
+                         "greater than 2: {}".format(bad))
+
+    shapes = tuple(a.shape for a in arrs)
+    shape = tuple(sum(x) for x in zip(*shapes))
+    dtype = cupy.find_common_type([a.dtype for a in arrs], [])
+    out = cupy.zeros(shape, dtype=dtype)
+    r, c = 0, 0
+    for arr in arrs:
+        rr, cc = arr.shape
+        out[r:r + rr, c:c + cc] = arr
+        r += rr
+        c += cc
+    return out
+
+
+def companion(a):
+    """Create a companion matrix.
+
+    Create the companion matrix associated with the polynomial whose
+    coefficients are given in ``a``.
+
+    Args:
+        a (cupy.ndarray): 1-D array of polynomial coefficients. The length of
+            ``a`` must be at least two, and ``a[0]`` must not be zero.
+
+    Returns:
+        (cupy.ndarray): The first row of the output is ``-a[1:]/a[0]``, and the
+            first sub-diagonal is all ones. The data-type of the array is the
+            same as the data-type of ``-a[1:]/a[0]``.
+
+    .. seealso:: :func:`cupyx.scipy.linalg.fielder_companion`
+    .. seealso:: :func:`scipy.linalg.companion`
+    """
+    n = a.size
+    if a.ndim != 1:
+        raise ValueError("`a` must be one-dimensional.")
+    if n < 2:
+        raise ValueError("The length of `a` must be at least 2.")
+    if a[0] == 0:
+        raise ValueError("The first coefficient in `a` must not be zero.")
```

Comment:
This check requires a device-to-host synchronization. CuPy does not raise an error in such a situation. For example, `cupy.linalg.inv` does not raise `LinAlgError` for singular matrix input.